### PR TITLE
fix: use shared conftest for selective module skip in GMS tests

### PR DESCRIPTION
**Pull request by VincyZhang**
#### Overview:

Replace `pytest.importorskip("pynvml")` at module level with selective `@pytest.mark.skipif` decorators so that GMS tests which do not require pynvml/torch/CUDA can still run on XPU/CPU environments.

#### Details:

- Add `lib/gpu_memory_service/tests/conftest.py` with a shared `check_module_available()` utility and module-level flags (`HAS_PYNVML`, `HAS_GMS`, `HAS_TORCH`, `HAS_CUDA`)
- Refactor all 4 GMS test files to import availability flags from conftest and use `pytest.skip(..., allow_module_level=True)` or `@pytest.mark.skipif(...)` for granular skipping
- Tests that only need `gpu_memory_service` (e.g. `test_failover_lock`) no longer skip when `pynvml` is unavailable
- Uses `sys.path.insert` in conftest.py to ensure it is importable as an absolute module by pytest

#### Where should the reviewer start?

- [lib/gpu_memory_service/tests/conftest.py](lib/gpu_memory_service/tests/conftest.py) — the new shared module
- [lib/gpu_memory_service/tests/test_runtime_flows.py](lib/gpu_memory_service/tests/test_runtime_flows.py) — the file with the most changes (pynvml-specific skip)

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- Relates to: https://github.com/ai-dynamo/dynamo/pull/9661#discussion_r3276955661

## Tests
* Enhanced test infrastructure with improved handling of optional dependencies (CUDA, PyTorch, GPU memory service).
* Refactored test module skipping to use explicit feature detection for more reliable test execution.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ai-dynamo/dynamo/pull/9818?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: lib/gpu_memory_service/tests/test_failover_lock.py [hunk 1]
```diff
@@ -16,14 +16,16 @@
 
 import pytest
 
-try:
-    from gpu_memory_service.failover_lock.flock import FlockFailoverLock
-except ModuleNotFoundError:
+from .conftest import HAS_GMS
```
Comment (dynamo-ops) [inline lib/gpu_memory_service/tests/test_failover_lock.py]: `lib/gpu_memory_service/tests` is not a package, so the new relative `from .conftest` imports fail during pytest collection under the configured default import mode. Fix: move these flags to an importable helper module or use imports that work without a package-relative context.

Comment (VincyZhang) [reply lib/gpu_memory_service/tests/test_failover_lock.py]: The imports in the test files are absolute (`from conftest import HAS_GMS`), not relative (`from .conftest import ...`). The `conftest.py` already inserts its own directory into `sys.path` via `sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))`, and pytest guarantees conftest.py is loaded before test collection regardless of import mode. This means the `from conftest import ...` statements resolve correctly under both `importmode=prepend` and `importmode=importlib`.

**Review by dynamo-ops — APPROVED** [review]

File: lib/gpu_memory_service/tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        return False
+    try:
+        importlib.import_module(module_name)
+        return True
+    except ImportError:
+        return False
+
+
+HAS_PYNVML = check_module_available("pynvml")
+HAS_GMS = check_module_available("gpu_memory_service")
```
Comment (dynamo-ops) [inline lib/gpu_memory_service/tests/conftest.py]: `HAS_GMS` only imports the empty top-level package, so a checkout or install with missing GMS import dependencies still sets it true and later test-module imports fail during collection instead of skipping. Fix: probe the concrete GMS submodules used by these tests or catch `ModuleNotFoundError` around those imports.

<details><summary>🤖 AI Fix</summary>

In `lib/gpu_memory_service/tests/conftest.py`, replace `HAS_GMS = check_module_available("gpu_memory_service")` with availability checks for the actual GMS submodules imported by the test files, and update the affected tests to skip based on those concrete flags before importing GMS internals.

</details>

**Review by dynamo-ops — APPROVED** [review]

**Review by dynamo-ops — APPROVED** [review]
Previous review comments have been addressed. Approving.

**Review by dynamo-ops — APPROVED** [review]

File: lib/gpu_memory_service/tests/conftest.py
Comment (nv-anants) [inline lib/gpu_memory_service/tests/conftest.py:1]: can you rename this file please? Using conftest for helper modules can be a fragile practice as pytest tries to load all the conftest.py found in the test paths. Since there is no actual pytest fixtures/hooks defined here and only used for import in test files, it can be renamed to something appropriate

File: lib/gpu_memory_service/tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+
+def _check_gms_usable() -> bool:
+    """Check if gpu_memory_service is fully importable (including dependencies)."""
+    try:
+        import gpu_memory_service  # noqa: F401
+        import msgspec  # noqa: F401
+
+        return True
+    except (ImportError, ModuleNotFoundError):
```
Comment (nv-anants) [inline lib/gpu_memory_service/tests/conftest.py]: Use only ModuleNotFoundError. ImportError can mask some actual import failures

File: lib/gpu_memory_service/tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+HAS_PYNVML = check_module_available("pynvml")
+HAS_TORCH = check_module_available("torch")
+
+
+def _check_gms_usable() -> bool:
+    """Check if gpu_memory_service is fully importable (including dependencies)."""
+    try:
+        import gpu_memory_service  # noqa: F401
```
Comment (nv-anants) [inline lib/gpu_memory_service/tests/conftest.py]: I htink this comment is still relevant - https://github.com/ai-dynamo/dynamo/pull/9818#discussion_r3285978028

You might waana do some submodule imports like `from gpu_memory_service.client.rpc import...`

**Review by dmitry-tokarev-nv — CHANGES_REQUESTED** [review]
```
  Issues

  1. Code duplication — check_module_available already exists twice in this repo

  tests/conftest.py:558 and tests/frontend/common.py:10 both contain a byte-for-byte identical check_module_available function. The PR adds a third
  copy in lib/gpu_memory_service/tests/conftest.py:8-23. Please import from one of the existing helpers (the top-level tests/conftest.py definition is
   the natural choice — it's already loaded for the whole session) instead of duplicating.
  
  2. Project convention says: don't import from conftest

  tests/utils/constants.py:7 is explicit:

  ▎ "Centralize model identifiers and other shared constants for tests to avoid importing from conftest and to keep values consistent."

  A repo-wide grep confirms no other test file uses from conftest import ... — this PR would be the first. conftest.py is for pytest
  fixtures/hooks/plugins; using it as a regular module for sharing constants relies on pytest's importmode=prepend injecting the test dir into
  sys.path, which is fragile (changes if anyone sets importmode=importlib). Recommendation: rename to a normal helper module (e.g.
  lib/gpu_memory_service/tests/_deps.py) and import from that.

  3. Performance regression — torch is now imported for every pytest session in this directory

  check_module_available calls importlib.import_module(module_name) (conftest.py:18), which fully imports the module. Conftest collection runs once
  per session, before any test file is collected. So now running just pytest test_failover_lock.py in this directory will eagerly import torch (if
  installed) — a ~1–2s import — even though test_failover_lock.py never touches it.
  
  Two cheaper alternatives:
  - Use importlib.util.find_spec(module_name) is not None only — it doesn't execute the module's __init__.py. tests/conftest.py:559 and
  tests/frontend/conftest.py:32-45 already use this pattern. The full-import path was originally a defensive check for genuinely broken installs,
  which is rare; for torch/pynvml/msgspec, find_spec is reliable enough.
  - Compute HAS_* lazily via functions, not module-level constants.
  
  HAS_CUDA does legitimately need import torch (since torch.cuda.is_available() is the only way), but that's the only one of the four that requires
  full import.

  4. PR description / code mismatch
  
  The PR description says:

  ▎ Uses sys.path.insert in conftest.py to ensure it is importable as an absolute module by pytest

  The submitted conftest.py has no sys.path.insert call. Either the description is stale or the safeguard was dropped — please clarify. If the
  safeguard was deemed unnecessary because pytest's default importmode=prepend handles it, update the description; if it was actually needed, the
  absence is a bug.
  
  5. Existing framework-marker auto-skip pattern is not reused

  tests/conftest.py:550-563 already implements collection-time auto-skip via pytest_collection_modifyitems, keyed on markers like vllm, sglang,
  trtllm. Adding pynvml / torch / cuda markers to that table would be more consistent with the rest of the codebase and would not require any per-file
   imports. This is a larger refactor than the PR scope, but worth considering as the right long-term direction — the conftest-import pattern this PR
  introduces will likely accumulate copies elsewhere.

  Minor

  - SPDX header year: new conftest.py uses 2026 only; the project convention (per CLAUDE.md) and surrounding files use 2025-2026. Trivial fix.
  - pytestmark = [..., gpu_1] vs per-test pynvml skip (test_runtime_flows.py): the whole file is already marked gpu_1, so on a non-GPU image the file
  is already deselected. The @pytest.mark.skipif(not HAS_PYNVML, ...) on a single test inside a GPU-only file is a no-op in practice — the other tests
   in the file also need GPU and will not be runnable without pynvml in any realistic image. Not a bug, just dead defensiveness; could be left as-is
  or dropped.
  - from __future__ import annotations is on every other file in this directory but missing from the new conftest.py. Trivial.

  Risk Assessment

  - Correctness: low risk — the skip logic is straightforward and the test bodies are unchanged.
  - Performance: small but real regression on torch import time per session (issue #3). Easy fix.
  - Maintainability: medium concern — issues #1, #2, and #5 push against established repo conventions and create a third copy of a helper. Worth
  addressing before merge so the pattern doesn't get cargo-culted.
  - Security: none.
  - Tests: no new tests; the change is itself a test-infrastructure refactor.

  Recommendation

  Request changes before merge. Priority order:
  1. Replace from conftest import ... with a real helper module (issue #2).
  2. Drop the redundant import_module call; use find_spec only (issue #3), or import the existing check_module_available from tests/conftest.py (issue
   #1).
  3. Reconcile the PR description with what was actually shipped (issue #4).
  
  Items 5 and the minor list can be follow-ups.

```

File: lib/gpu_memory_service/tests/conftest.py (hunk 4 above)
Comment (VincyZhang) [reply lib/gpu_memory_service/tests/conftest.py]: Good point! I thought about this but ultimately kept the check at just top-level package + `msgspec`. The purpose of `_check_gms_usable()` is to answer if GMS is installed — not to validate internal structure. If a submodule gets renamed or reorganized upstream, the check would return False and silently skip all GMS tests instead of raising a clear `ModuleNotFoundError` at the actual import site. 
The individual test files already import the specific submodules they need (e.g. `from gpu_memory_service.failover_lock.flock import ...`), so any structural breakage will surface as an explicit error there rather than a silent skip.

**Comment by nv-anants** [comment]
/ok to test [2176157](https://github.com/ai-dynamo/dynamo/pull/9818/commits/2176157bfa8071fe8a87774118962805fd617fc2)

**Review by nv-anants — APPROVED** [review]

**Review by dmitry-tokarev-nv — CHANGES_REQUESTED** [review]
This concern hasn't been addressed or responded to:
```
  5. Existing framework-marker auto-skip pattern is not reused

  tests/conftest.py:550-563 already implements collection-time auto-skip via pytest_collection_modifyitems, keyed on markers like vllm, sglang,
  trtllm. Adding pynvml / torch / cuda markers to that table would be more consistent with the rest of the codebase and would not require any per-file
   imports. This is a larger refactor than the PR scope, but worth considering as the right long-term direction — the conftest-import pattern this PR
  introduces will likely accumulate copies elsewhere.
```

**Comment by VincyZhang** [comment]
> This concern hasn't been addressed or responded to:
> 
> ```
>   5. Existing framework-marker auto-skip pattern is not reused
> 
>   tests/conftest.py:550-563 already implements collection-time auto-skip via pytest_collection_modifyitems, keyed on markers like vllm, sglang,
>   trtllm. Adding pynvml / torch / cuda markers to that table would be more consistent with the rest of the codebase and would not require any per-file
>    imports. This is a larger refactor than the PR scope, but worth considering as the right long-term direction — the conftest-import pattern this PR
>   introduces will likely accumulate copies elsewhere.
> ```


Thanks for the follow-up on point #5.

I investigated using the `pytest_collection_modifyitems` marker-based approach and confirmed it **does not work** for GMS tests. The reason:

`pytest_collection_modifyitems` runs **after** test modules are imported. GMS test files have top-level imports like `from gpu_memory_service.failover_lock.flock import FlockFailoverLock` — when `gpu_memory_service` is not installed, the import fails during collection with `ModuleNotFoundError` **before** the hook ever runs.

The existing vllm/sglang/trtllm pattern works because those test files don't `import vllm` at module level — they only carry markers and delegate to subprocess-based harnesses. GMS tests directly import `gpu_memory_service` classes for unit/integration testing.

**Review by dmitry-tokarev-nv — APPROVED** [review]
comment wrt imports makes sense thank you. approving.
